compositor_pipeline/src/pipeline/input/mp4/mp4_file_reader.rs

```diff
@@ -0,0 +1,369 @@
+        stop_thread: Arc<AtomicBool>,
+        input_id: InputId,
+    ) -> Result<Option<(Self, ChunkReceiver)>, Mp4Error> {
+        let reader = mp4::Mp4Reader::read_header(reader, size)?;
+
+        let Some(track_info) = track_info_reader(&reader) else {
+            return Ok(None);
+        };
+
+        let (sender, receiver) = crossbeam_channel::bounded(10);
+
+        let stop_thread_clone = stop_thread.clone();
+        let decoder_options = track_info.decoder_options.clone();
+
+        std::thread::Builder::new()
+            .name(format!("mp4 reader {input_id}"))
+            .spawn(move || {
+                let _span = span!(Level::INFO, "MP4", input_id = input_id.to_string()).entered();
+                run_reader_thread(
+                    reader,
+                    sender,
+                    stop_thread_clone,
+                    fragment_receiver,
+                    track_info,
+                );
+                debug!("Closing MP4 reader thread");
+            })
+            .unwrap();
+
+        Ok(Some((
+            Mp4FileReader {
+                stop_thread,
+                fragment_sender: None,
+                decoder_options,
+            },
+            receiver,
+        )))
+    }
+
+    pub fn decoder_options(&self) -> DecoderOptions {
+        self.decoder_options.clone()
+    }
+}
+
+impl<D> Drop for Mp4FileReader<D> {
+    fn drop(&mut self) {
+        self.stop_thread
+            .store(true, std::sync::atomic::Ordering::Relaxed)
+    }
+}
+
+fn run_reader_thread<Reader: Read + Seek, DecoderOptions>(
+    mut reader: Mp4Reader<Reader>,
+    sender: Sender<PipelineEvent<EncodedChunk>>,
+    stop_thread: Arc<AtomicBool>,
+    _fragment_receiver: Option<Receiver<PipelineEvent<Bytes>>>,
+    track_info: TrackInfo<DecoderOptions, impl FnMut(mp4::Mp4Sample) -> Bytes>,
+) {
+    let mut sample_unpacker = track_info.sample_unpacker;
+
+    'samples: for i in 1..track_info.sample_count {
+        match reader.read_sample(track_info.track_id, i) {
+            Ok(Some(sample)) => {
+                let rendering_offset = sample.rendering_offset;
+                let start_time = sample.start_time;
+                let data = sample_unpacker(sample);
+
+                let dts = Duration::from_secs_f64(start_time as f64 / track_info.timescale as f64);
+                let chunk = EncodedChunk {
+                    data,
+                    pts: Duration::from_secs_f64(
+                        (start_time as f64 + rendering_offset as f64) / track_info.timescale as f64,
+                    ),
+                    dts: Some(dts),
+                    kind: track_info.chunk_kind,
+                };
+
+                let mut chunk = Some(PipelineEvent::Data(chunk));
+                loop {
+                    match sender.send_timeout(chunk.take().unwrap(), Duration::from_millis(50)) {
+                        Ok(()) => {
+                            continue 'samples;
```

Comment:
if you need labels then it's a sign to refactor, especially here. A function that tries to write to a channel while periodically checking for should_close could be useful in few places.